oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentialSource.java

```diff
@@ -48,6 +50,157 @@ public class IdentityPoolCredentialSource extends ExternalAccountCredentials.Cre
+   * @return A new {@link CertificateConfig} instance.
+   * @throws IllegalArgumentException if the 'certificate' entry is not a Map or if required fields
+   *     within the certificate configuration have invalid types.
+   */
+  private CertificateConfig getCertificateConfig(Map<String, Object> credentialSourceMap) {
+    Object certValue = credentialSourceMap.get("certificate");
+    if (!(certValue instanceof Map)) {
+      throw new IllegalArgumentException(
+          "The 'certificate' credential source must be a JSON object (Map).");
+    }
+    Map<String, Object> certificateMap = (Map<String, Object>) certValue;
+
+    Boolean useDefaultCertificateConfig =
+        getOptionalBoolean(certificateMap, "use_default_certificate_config");
+    String trustChain = getOptionalString(certificateMap, "trust_chain_path");
+    String certificateConfigLocation =
+        getOptionalString(certificateMap, "certificate_config_location");
+
+    return new CertificateConfig(
+        useDefaultCertificateConfig, certificateConfigLocation, trustChain);
+  }
+
+  /**
+   * Retrieves an optional boolean value from a map.
+   *
+   * @param map The map to retrieve from.
+   * @param key The key of the boolean value.
+   * @return The boolean value if present and of the correct type, otherwise null.
+   * @throws IllegalArgumentException if the value is present but not a boolean.
+   */
+  private @Nullable Boolean getOptionalBoolean(Map<String, Object> map, String key) {
+    Object value = map.get(key);
+    if (value == null) {
+      return null;
+    }
+    if (!(value instanceof Boolean)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Invalid type for '%s' in certificate configuration: expected Boolean, got %s.",
+              key, value.getClass().getSimpleName()));
+    }
+    return (Boolean) value;
+  }
+
+  /**
+   * Retrieves an optional string value from a map.
+   *
+   * @param map The map to retrieve from.
+   * @param key The key of the string value.
+   * @return The string value if present and of the correct type, otherwise null.
+   * @throws IllegalArgumentException if the value is present but not a string.
+   */
+  private @Nullable String getOptionalString(Map<String, Object> map, String key) {
+    Object value = map.get(key);
+    if (value == null) {
+      return null;
+    }
+    if (!(value instanceof String)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Invalid type for '%s' in certificate configuration: expected String, got %s.",
+              key, value.getClass().getSimpleName()));
+    }
+    return (String) value;
+  }
+  /**
+   * Represents the configuration options for X.509-based workload credentials (mTLS). It specifies
+   * how to locate and use the client certificate, private key, and optional trust chain for mutual
+   * TLS authentication.
+   */
+  public static class CertificateConfig implements java.io.Serializable {
```

Comment:
qq, why does this class need to be Serializable?

Reply:
Good question! CertificateConfig is marked as Serializable because it's used as a field within IdentityPoolCredentialSource. Since IdentityPoolCredentialSource itself needs to be Serializable (as it extends ExternalAccountCredentials.CredentialSource), my understanding is that any non-transient fields it holds, like the certificateConfig instance, must also implement Serializable to allow the containing object to be serialized correctly.